packages/core/src/chopsticks-provider.ts

```diff
@@ -1,286 +1,245 @@
-//             return
-//           }
-//           sub?.onCancel?.()
-//           delete this.#subscriptions[e.data.subid]
-//         }
-//         break
-
-//       case 'send-result':
-//         {
-//           const handler = this.#handlers[e.data.id]
-//           if (!handler) {
-//             logger.error(`Unable to find handler=${e.data.id}`)
-//             return
-//           }
-//           logger.debug('send-result', {
-//             method: e.data.method,
-//             result: JSON.parse(e.data.result || '{}'),
-//             data: e.data,
-//           })
-//           try {
-//             handler.callback(null, e.data.result ? JSON.parse(e.data.result) : undefined)
-//           } catch (error) {
-//             handler.callback(error as Error, undefined)
-//           }
-//           delete this.#handlers[e.data.id]
-//         }
-//         break
-
-//       default:
-//         break
-//     }
-//   }
-// }
+import { EventEmitter } from 'eventemitter3'
+import {
+  ProviderInterface,
+  ProviderInterfaceCallback,
+  ProviderInterfaceEmitCb,
+  ProviderInterfaceEmitted,
+  ProviderStats,
+} from '@polkadot/rpc-provider/types'
+
+import { Blockchain } from './blockchain'
+import { StorageValues, setStorage } from './utils'
+import { allHandlers } from './rpc'
+import { defaultLogger } from './logger'
+import { setup } from './setup'
+
+const providerHandlers = {
+  ...allHandlers,
+  new_block: async (context: any, _params: any, _subscriptionManager: any) => {
+    const { chain } = context
+    const block = await chain.newBlock()
+    return block
+  },
+}
+
+const logger = defaultLogger.child({ name: '[Chopsticks provider]' })
+
+interface SubscriptionHandler {
+  callback: ProviderInterfaceCallback
+  type: string
+}
+
+interface Subscription extends SubscriptionHandler {
+  method: string
+  params?: unknown[]
+  onCancel?: () => void
+  result?: unknown
+}
+
+export interface ChopsticksProviderProps {
+  /** upstream endpoint */
+  endpoint: string
+  /** default to latest block */
+  blockHash?: string
+  dbPath?: string
+  storageValues?: StorageValues
+}
+
+/**
+ * A provider for ApiPromise.
+ *
+ * Currectly only support browser environment.
+ */
+export class ChopsticksProvider implements ProviderInterface {
+  #isConnected = false
+  #eventemitter: EventEmitter
+  #isReadyPromise: Promise<void>
+  #endpoint: string
+  readonly stats?: ProviderStats
+  #subscriptions: Record<string, Subscription> = {}
+  #blockHash: string | undefined
+  #dbPath: string | undefined
+  #storageValues: StorageValues | undefined
+  #chain: Blockchain | undefined
+
+  constructor({ endpoint, blockHash, dbPath, storageValues }: ChopsticksProviderProps) {
+    if (!endpoint) {
+      throw new Error('ChopsticksProvider requires the upstream endpoint')
+    }
+    this.#endpoint = endpoint
+    this.#blockHash = blockHash
+    this.#dbPath = dbPath
+    this.#storageValues = storageValues
```

Comment:
No need to duplicate stuff here. Just make chain public so setStorage can be called outside after provider is ready